python/packages/autogen-agentchat/src/autogen_agentchat/teams/_group_chat/_base_group_chat_manager.py

```diff
@@ -140,58 +142,65 @@ async def handle_start(self, message: GroupChatStart, ctx: MessageContext) -> No
+                    await self._signal_termination(stop_message)
+                    # Stop the group chat.
+                    return
 
-        # Select a speaker to continue the conversation.
-        speaker_name_future = asyncio.ensure_future(self.select_speaker(self._message_thread))
-        # Link the select speaker future to the cancellation token.
-        ctx.cancellation_token.link_future(speaker_name_future)
-        speaker_name = await speaker_name_future
-        if speaker_name not in self._participant_name_to_topic_type:
-            raise RuntimeError(f"Speaker {speaker_name} not found in participant names.")
-        speaker_topic_type = self._participant_name_to_topic_type[speaker_name]
-        await self.publish_message(
-            GroupChatRequestPublish(),
-            topic_id=DefaultTopicId(type=speaker_topic_type),
-            cancellation_token=ctx.cancellation_token,
-        )
+            # Increment the turn count.
+            self._current_turn += 1
+            # Check if the maximum number of turns has been reached.
+            if self._max_turns is not None:
+                if self._current_turn >= self._max_turns:
+                    stop_message = StopMessage(
+                        content=f"Maximum number of turns {self._max_turns} reached.",
+                        source=self._name,
+                    )
+                    # Reset the termination conditions and turn count.
+                    if self._termination_condition is not None:
+                        await self._termination_condition.reset()
+                    self._current_turn = 0
+                    # Signal termination to the caller of the team.
+                    await self._signal_termination(stop_message)
+                    # Stop the group chat.
+                    return
+
+            # Select a speaker to continue the conversation.
+            speaker_name_future = asyncio.ensure_future(self.select_speaker(self._message_thread))
+            # Link the select speaker future to the cancellation token.
+            ctx.cancellation_token.link_future(speaker_name_future)
+            speaker_name = await speaker_name_future
+            if speaker_name not in self._participant_name_to_topic_type:
+                raise RuntimeError(f"Speaker {speaker_name} not found in participant names.")
+            speaker_topic_type = self._participant_name_to_topic_type[speaker_name]
+            await self.publish_message(
+                GroupChatRequestPublish(),
+                topic_id=DefaultTopicId(type=speaker_topic_type),
+                cancellation_token=ctx.cancellation_token,
+            )
+        except Exception as e:
+            # Handle the exception and signal termination with an error.
+            error = SerializableException.from_exception(e)
+            await self._signal_termination_with_error(error)
+            # Raise the exception to the runtime.
+            raise e
```

Comment:
this should just be `raise` so we can preserve the original exception's stack trace